code/game/objects/items/weapons/tools/_tools.dm

```diff
@@ -17,13 +17,66 @@
 	var/passive_fuel_cost = 0.03 //Fuel consumed per process tick while active
 	var/max_fuel = 0
 
+	//Third type of resource, stock. A tool that uses physical objects (or itself) in order to work
+	//Currently used for tape roll
+	var/use_stock_cost = 0
+	var/stock = 0
+	var/max_stock = 0
+	var/allow_decimal_stock = TRUE
+	var/delete_when_empty = TRUE
+
 	var/toggleable = FALSE	//Determinze if it can be switched ON or OFF, for example, if you need a tool that will consume power/fuel upon turning it ON only. Such as welder.
 	var/switched_on = FALSE	//Curent status of tool. Dont edit this in subtypes vars, its for procs only.
 	var/switched_on_qualities = null	//This var will REPLACE tool_qualities when tool will be toggled on.
 	var/switched_off_qualities = null	//This var will REPLACE tool_qualities when tool will be toggled off. So its possible for tool to have diferent qualities both for ON and OFF state.
 	var/create_hot_spot = FALSE	//Set this TRUE to ignite plasma on turf with tool upon activation
 	var/glow_color = null	//Set color of glow upon activation, or leave it null if you dont want any light
 
+
+/******************************
+	/* Core Procs */
+*******************************/
+//Fuel and cell spawn
+/obj/item/weapon/tool/New()
+	..()
+	if(!cell && suitable_cell)
+		cell = new suitable_cell(src)
+
+	if(use_fuel_cost)
+		var/datum/reagents/R = new/datum/reagents(max_fuel)
+		reagents = R
+		R.my_atom = src
+		R.add_reagent("fuel", max_fuel)
+
+	if (use_stock_cost)
+		stock = max_stock
+
+	update_icon()
+	return
+
+//For killing processes like hot spots
+/obj/item/weapon/tool/Destroy()
+	if (src in SSobj.processing)
+		STOP_PROCESSING(SSobj, src)
+	return ..()
+
+//Ignite plasma around, if we need it
+/obj/item/weapon/tool/Process()
+	if(switched_on)
+		if(create_hot_spot)
+			var/turf/location = src.loc
```

Comment:
Just use `get_turf(src)` for this whole section, you don't need any intricate location checking.
`#define get_turf(atom) get_step(atom, 0)`